test/test_ops.py

```diff
@@ -38,6 +43,89 @@ def test_unsupported_dtypes(self, device, dtype, op):
+                             msg,
+                             exact_device=False,
+                             **kwargs)
+        else:
+            self.assertEqual(actual, expected, msg, exact_device=False, **kwargs)
+
+    def _test_reference_numerics(self, dtype, op, tensors, equal_nan=True):
+        def _helper_reference_numerics(expected, actual, msg, exact_dtype, equal_nan=True):
+            if not torch.can_cast(numpy_to_torch_dtype_dict[expected.dtype.type], dtype):
+                exact_dtype = False
+
+            if dtype in [torch.uint8, torch.int8, torch.bool]:
+                self.assertEqualHelper(actual, expected, msg, dtype=dtype,
+                                       exact_dtype=exact_dtype, rtol=1e-3, atol=1e-2)
+
+            elif dtype is torch.bfloat16:
+                self.assertEqualHelper(actual, expected, msg, dtype=dtype,
+                                       exact_dtype=exact_dtype, rtol=16e-3, atol=1e-5)
+
+            else:
+                self.assertEqualHelper(actual, expected, msg, dtype=dtype, equal_nan=equal_nan, exact_dtype=exact_dtype)
+
+        for sample_input in tensors:
+            sample_args = sample_input.args
+            sample_kwargs = sample_input.kwargs
+            sample_input = sample_input.input
+            torch_kwargs, numpy_kwargs = op.sample_kwargs(sample_input.device, dtype, sample_input)
+            numpy_args = []
+            if dtype is torch.bfloat16:
+                for sample_arg in sample_args:
+                    numpy_arg = sample_arg.cpu().to(torch.float32).numpy() if isinstance(sample_arg, torch.Tensor) \
+                                    else sample_arg
+                    numpy_args.append(numpy_arg)
+                sample_numpy_input = sample_input.cpu().clone().to(torch.float32).numpy()
+            else:
+                for sample_arg in sample_args:
+                    numpy_arg = sample_arg.cpu().numpy() if isinstance(sample_arg, torch.Tensor) else sample_arg
+                    numpy_args.append(numpy_arg)
+                sample_numpy_input = sample_input.cpu().clone().numpy()
+            actual = op(sample_input, *sample_args, **torch_kwargs)
+            expected = op.ref(sample_numpy_input, *numpy_args, **numpy_kwargs)
+
+            if sample_input.numel() < 10:
+                msg = ("Failed to produce expected results! Input tensor was"
+                        " {0}, torch result is {1}, and reference result is"
+                        " {2}.").format(sample_input, actual, expected)
+            else:
+                msg = None
+
+            exact_dtype = True
+            if isinstance(actual, torch.Tensor):
+                _helper_reference_numerics(expected, actual, msg, exact_dtype, equal_nan)
+            else:
+                for x, y in zip(expected, actual):
+                    _helper_reference_numerics(x, y, msg, exact_dtype, equal_nan)
+
+    @suppress_warnings
```

Comment:
Add a comment for this test

Reply:
Yes, apologies. I'll add a comment here.